frame/support/src/storage/stream_iter.rs

```diff
@@ -0,0 +1,405 @@
+		T::stream_iter(Self::hashed_key().into())
+	}
+}
+
+/// A streaming iterator implementation for SCALE container types.
+///
+/// SCALE container types follow the same type of encoding `Compact<u32>(len) ++ data`.
+/// This type provides an [`Iterator`](sp_std::iter::Iterator) implementation that decodes
+/// one item after another with each call to [`next`](Self::next). The bytes representing
+/// the container are also not read at once into memory and instead being read in chunks. As long
+/// as individual items are smaller than these chunks the memory usage of this iterator should
+/// be constant. On decoding errors [`next`](Self::next) will return `None` to signal that the
+/// iterator is finished.
+pub struct ScaleContainerStreamIter<T> {
+	marker: sp_std::marker::PhantomData<T>,
+	input: StorageInput,
+	length: u32,
+	read: u32,
+}
+
+impl<T> ScaleContainerStreamIter<T> {
+	/// Creates a new instance of the stream iterator.
+	///
+	/// - `key`: Storage key of the container in the state.
+	pub fn new(key: Vec<u8>) -> Self {
+		let mut input = StorageInput::new(key);
+		let length = if input.exists() {
+			match codec::Compact::<u32>::decode(&mut input) {
+				Ok(length) => length.0,
+				Err(e) => {
+					// TODO #3700: error should be handleable.
+					log::error!(
+						target: "runtime::storage",
+						"Corrupted state at `{:?}`: {:?}",
+						input.key,
+						e,
+					);
+
+					0
+				},
+			}
+		} else {
+			0
+		};
+
+		Self { marker: sp_std::marker::PhantomData, input, length, read: 0 }
+	}
+}
+
+impl<T: codec::Decode> sp_std::iter::Iterator for ScaleContainerStreamIter<T> {
+	type Item = T;
+
+	fn next(&mut self) -> Option<T> {
+		if self.read >= self.length {
+			None
+		} else {
+			self.read += 1;
+
+			codec::Decode::decode(&mut self.input).ok()
+		}
+	}
+}
+
+/// The size of the internal buffer used by [`StorageInput`].
+///
+/// This internal buffer is used to speed up implementation as reading from the
+/// state for every access is too slow.
+const STORAGE_INPUT_BUFFER_CAPACITY: usize = 2 * 1024;
+
+/// Implementation of [`codec::Input`] using [`sp_io::storage::read`].
+///
+/// Keeps an internal buffer with a size of [`STORAGE_INPUT_BUFFER_CAPACITY`]. All read accesses
+/// are tried to be served by this buffer. If the buffer doesn't hold enough bytes to fullfill the
+/// current read access, the buffer is re-filled from the state. A read request that is bigger than
+/// the internal buffer is directly forwarded to the state to reduce the number of reads from the
+/// state.
+struct StorageInput {
+	key: Vec<u8>,
+	offset: u32,
+	total_length: u32,
+	exists: bool,
+	buffer: Vec<u8>,
+	buffer_pos: usize,
+}
+
+impl StorageInput {
+	/// Create a new instance of the input.
+	///
+	/// - `key`: The storage key of the storage item that this input will read.
+	fn new(key: Vec<u8>) -> Self {
+		let mut buffer = Vec::with_capacity(STORAGE_INPUT_BUFFER_CAPACITY);
+		unsafe {
+			buffer.set_len(buffer.capacity());
+		}
```

Comment:
*Technically* you're right, but in practice this issue's kinda tricky.

Doing this "properly" would require us to zero-initialize the buffer (which is kinda silly; it's a waste of CPU in this case for no practical benefit) *or* access the buffer only through a pointer.

And it's also up to debate how you define "initialized". We control our allocator. When running within WASM (at this point I don't think we have to care about the native executor; it's on its way out anyway) technically it does always return properly initialized `u8`s, even if their contents might be unspecified if same memory is reused. (The [docs](https://doc.rust-lang.org/std/alloc/trait.GlobalAlloc.html#tymethod.alloc) explicitly say that the memory "may or may not be initialized", so it does leave the possibility for it to *be* initialized.) So does accessing such memory trigger UB? Does creating a slice to such memory trigger UB? Nobody knows. But in practice, *right now*, I don't think it does. (And `miri` agrees with me if you write a simple program that does something like this. Also, @bkchr just to be sure, could you later run the tests for this under `cargo miri test` just in case?)

If we'd want this to be completely proper we'd have to add an extra function to read the storage through a raw pointer. (Which technically AFAIK shouldn't require a new host function; just a different interface to the *existing* one.) Then you'd only `set_len` it *after* reading from storage. Not sure if it's actually worth it for a technicality like this, but it is a possibility.